### ISSUE TITLE
Web hook notification sends wrong Authorization header 

### ISSUE THREAD

**Issue by larsbergstrom** [travis-ci/travis-ci#6435]
I've had a setup with authorized web hook notifications for https://github.com/servo/rust-mozjs (and many other repos) working for quite some time, but the checksum sent from Travis appears to be incorrect since 2016-08-04.

We are currently using the @bors-servo user and token for the travis build. This has been working for at least a year.

This is currently breaking our autolander and is requiring manual intervention on the ~20 repositories that we normally have a robot autoland!

I've verified that the the Authorization header sent from Travis does not match what the documentation at http://about.travis-ci.org/docs/user/notifications/#Authorization suggest, and the code I'm using in the receiving service (which is unchanged):
https://github.com/servo/homu/blob/master/homu/server.py#L617-L627

More info on our debugging is here: https://github.com/servo/servo/issues/12739

Looks similar to: https://github.com/travis-ci/travis-ci/issues/1262

CC: @BanzaiMan @mkubenka @edunham

**Comment by larsbergstrom** [comment]
From what I can tell, this code appears to live in this file:
https://github.com/travis-ci/travis-tasks/blob/master/lib/travis/addons/webhook/task.rb

And it looks like at least since 2014 this code has been as described in the docs, that is:
https://github.com/travis-ci/travis-tasks/blob/master/lib/travis/addons/webhook/task.rb#L53
```ruby
Digest::SHA2.hexdigest(repo_slug + params[:token].to_s)
```

Where:
https://github.com/travis-ci/travis-tasks/blob/master/lib/travis/addons/webhook/task.rb#L65
```ruby
def repo_slug
            repository.values_at(:owner_name, :name).join('/')
end
```

The only other thing that I can imagine is that we're somehow accidentally now taking the other code path here:
https://github.com/travis-ci/travis-tasks/blob/master/lib/travis/addons/webhook/task.rb#L37-L41

And so now the `'Authorization'` header is the user & password? Hrm. Most of the relevant code hasn't been changed in a long, long time, though.

**Comment by mkubenka** [comment]
I did some testing and it looks like `Authorization` header is just `repo_slug` (_owner_name/repo_name_). So, maybe there were some change with token and it's not correctly populated from DB...

**Comment by larsbergstrom** [comment]
Wow, I just confirmed this is the case! Nice sleuthing :-)

**Comment by mkubenka** [comment]
Any update on this?

**Comment by larsbergstrom** [comment]
@mkubenka It sounded like a full fix for this was at least a few weeks out, so we switched to instead checking the results of `continuous-integration/travis-ci/push` in the GitHub status notifications for that repository. Even if the tests are not to a PR branch (as in our case, where we push to an `auto` branch for our CI tests), we still get the results there and can act on them.

**Comment by joshk** [comment]
Hi All

I wanted to give an update on what we are doing about this issue.

As a brief history, the use of the username and 'token' in the GitHub webhook settings is a relic of the past. The intention were good, in that we required some form of auth to hook up a repository to Travis, but the implementation had some faults as the token was linked to the user and not to the repo, and as users come and go over time (from having access to a repo, to not any more), it resulted in confusion.

On top of this, several years ago we removed the need to 'auth' a repo build request, and instead relied on first knowing if the repo existed (via sync), and confirming that the build request came from GitHub.

Before we removed the 'auth' bits in Gatekeeper, as mentioned above, when we came to implement webhooks in travis-tasks, some users asked for some form of auth and we thought we could reuse the token.

What we decided to do was add the repo slug and token together, and then do some hex digesting =>` Digest::SHA2.hexdigest(repo_slug + params[:token])` (https://github.com/travis-ci/travis-tasks/blob/79bf24bc485420730745aa78af957e80247882f8/lib/travis/addons/webhook/task.rb#L53-L56), which we would then put in the Authorization header.

With the token being hard to determine on a per repo basis, the initial idea was to move the token to being attached to the repo and not the user, and showing it on the settings page.

The great @rkh had a great idea, he suggested we create a Signature of the content using a Private key, and expose our Public key so people can verify it.

https://github.com/travis-ci/travis-tasks/pull/65

We are working on testing out this PR on staging, and then moving forward with deploying this to Production. We believe this to be the best way forward, and the preferred long term solution.

I will update this issue with more information as it becomes available.

**Comment by BanzaiMan** [comment]
Hi, everyone.

We have implemented the new method of verifying the webhook notification using the private key signing @joshk described earlier.

Here are the steps:

1. Pick up the `payload` data from the HTTP request's body.
1. Obtain the `Signature` header value.
1. Obtain the public key corresponding to the private key that signed     
   the payload. This is available at the `/config` endpoint's             
   `config.notifications.webhook.public_key` on the relevant API server (https://api.travis-ci.org or https://api.travis-ci.com).
1. Verify the signature using the public key and SHA1 digest.

I've opened a [documentation PR](https://github.com/travis-ci/docs-travis-ci-com/pull/710) to describe the steps in more detail.

Tentatively I put down November 1, 2016 as the date on which the current method will be deprecated.
